src/lib/logs/log_self_monitoring.py

```diff
@@ -0,0 +1,309 @@
+#     Copyright 2020 Dynatrace LLC
+#
+#     Licensed under the Apache License, Version 2.0 (the "License");
+#     you may not use this file except in compliance with the License.
+#     You may obtain a copy of the License at
+#
+#         http://www.apache.org/licenses/LICENSE-2.0
+#
+#     Unless required by applicable law or agreed to in writing, software
+#     distributed under the License is distributed on an "AS IS" BASIS,
+#     WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#     See the License for the specific language governing permissions and
+#     limitations under the License.
+import asyncio
+import os
+import queue
+import time
+import traceback
+from collections import Counter
+from queue import Queue
+from typing import Dict, List
+
+import aiohttp
+
+from lib.context import LoggingContext, LogsSfmContext, DynatraceConnectivity, LogsContext
+from lib.credentials import create_token, get_dynatrace_log_ingest_url_from_env
+from lib.instance_metadata import InstanceMetadata
+from lib.logs.log_forwarder_variables import SENDING_WORKER_EXECUTION_PERIOD_SECONDS, MAX_MESSAGES_PROCESSED, \
+    LOGS_SUBSCRIPTION_PROJECT, LOGS_SUBSCRIPTION_ID
+from lib.logs.log_sfm_metric_descriptor import LOG_SELF_MONITORING_CONNECTIVITY_METRIC_TYPE, \
+    LOG_SELF_MONITORING_ALL_REQUESTS_METRIC_TYPE, \
+    LOG_SELF_MONITORING_TOO_OLD_RECORDS_METRIC_TYPE, LOG_SELF_MONITORING_PARSING_ERRORS_METRIC_TYPE, \
+    LOG_SELF_MONITORING_PROCESSING_TIME_METRIC_TYPE, LOG_SELF_MONITORING_SENDING_TIME_SIZE_METRIC_TYPE, \
+    LOG_SELF_MONITORING_TOO_LONG_CONTENT_METRIC_TYPE, LOG_SELF_MONITORING_LOG_INGEST_PAYLOAD_SIZE_METRIC_TYPE, \
+    LOG_SELF_MONITORING_SENT_LOGS_ENTRIES_METRIC_TYPE
+from lib.logs.log_sfm_metrics import LogSelfMonitoring
+from lib.self_monitoring import push_self_monitoring_time_series
+
+
+def aggregate_self_monitoring_metrics(aggregated_sfm: LogSelfMonitoring, sfm_list: List[LogSelfMonitoring]):
+    for sfm in sfm_list:
+        aggregated_sfm.all_requests += sfm.all_requests
+        aggregated_sfm.too_old_records += sfm.too_old_records
+        aggregated_sfm.parsing_errors += sfm.parsing_errors
+        aggregated_sfm.records_with_too_long_content += sfm.records_with_too_long_content
+        aggregated_sfm.dynatrace_connectivity.extend(sfm.dynatrace_connectivity)
+        aggregated_sfm.processing_time += sfm.processing_time
+        aggregated_sfm.sending_time += sfm.sending_time
+        aggregated_sfm.log_ingest_payload_size += sfm.log_ingest_payload_size
+        aggregated_sfm.sent_logs_entries += sfm.sent_logs_entries
+    return aggregated_sfm
+
+
+def put_sfm_into_queue(context: LogsContext):
+    try:
+        context.sfm_queue.put(context.self_monitoring, False, SENDING_WORKER_EXECUTION_PERIOD_SECONDS + 1)
```

Comment:
(Minor) this unsused timeout value `SENDING_WORKER_EXECUTION_PERIOD_SECONDS + 1` is confusing at glance, you can use `put_nowait(item)` instead of `put` to be more explicit that this call is not blocking